### PR TITLE
Taylor factorial fix

### PR DESCRIPTION
**Pull request by SamDuffield**
Generalises `gaussian.taylor.init_prepare` to support mean and chol_cov extraction when the linearisation point has a factorial axis.

It's a bit annoying that this support is added in the non-factorial code but I couldn't see a nicer way to do it. It doesn't change the functionality when no factorial dim is provided or add any computation.

### CODE REVIEW

**Comment by Sahel13** [comment]
Just making sure, this assumes that `init_log_density` describes one factor, and that all factors have the same density function, so is that a restriction we're fine with making? Also, don't we have the same shape mismatch when calling `linearize_log_density` in `filter_combine`?

**Comment by SamDuffield** [comment]
Yeah it assumes all factors have the same density function. This is not ideal, so definitely worth thinking about but I think for a future PR (unless you have a good idea to support it? 😅).

We don’t have this problem in `filter_combine` as that gets called after a `join` where factors are stacked into a single state with no factorial axis.

**Comment by Sahel13** [comment]
> Yeah it assumes all factors have the same density function. This is not ideal, so definitely worth thinking about but I think for a future PR (unless you have a good idea to support it? 😅).

I think I might have a way, but we can definitely put that in a different PR.

> We don’t have this problem in filter_combine as that gets called after a join where factors are stacked into a single state with no factorial axis.

Yeah good point. What about the `jnp.cov` in `filter_prepare` though? That won't give the right shape with factorial inputs. I think we need to replace that with something like:
```py
dummy_chol_cov = dummy_tree_like(jnp.zeros(dummy_mean.shape + (dummy_mean.shape[-1],), dtype=dummy_mean.dtype,))
```

**Comment by SamDuffield** [comment]
> I think I might have a way, but we can definitely put that in a different PR.

👀 excited to see this



> Yeah good point. What about the `jnp.cov` in `filter_prepare` though? That won't give the right shape with factorial inputs. I think we need to replace that with something like:

I don’t think `filter_prepare` will ever see a state input with a factorial axis. Means should be shape `(num_factors * x_dim, )` and chol_covs `(num_factors * x_dim, num_factors * x_dim)` as this is what `factorial.gaussian.join` provides. So shouldn’t be a problem?

**Comment by Sahel13** [comment]
Uh `filter_prepare` gets its shape from the linearization point returned by `get_init_log_density`, right? So I'm worried about this part:
```py
    dummy_mean_struct = eval_shape(lambda mi: get_init_log_density(mi)[1], model_inputs)
    dummy_mean = dummy_tree_like(dummy_mean_struct)
    dummy_chol_cov = dummy_tree_like(jnp.cov(dummy_mean[..., None]))
```

**Comment by SamDuffield** [comment]
Ohhh yeah of course! That’ll need fixing too, dang

**Comment by SamDuffield** [comment]
I'm wondering if it be possible to infer the requires shapes from `get_dynamics_log_density` rather than `get_init_log_density`. That would solve this problem and also be more natural for `filter_prepare`

Edit: Nope I think this is a bad idea, as `get_dynamics_log_density` requires a `state` on input which `filter_prepare` doesn't have. And we want to support the linearization point being constructed from the input `state`

**Comment by SamDuffield** [comment]
@Sahel13 - I added the covariance shape handling. Are we good to merge? 👀

**Comment by Sahel13** [comment]
Yup this works!

**Review by Sahel13 — APPROVED** [review]

**Review by Sahel13 — CHANGES_REQUESTED** [review]
You forgot to change this for taylor XD

**Comment by SamDuffield** [comment]
What a moron, should have used AI, they'd never make such a foolish mistake lol

Fixed now?

**Review by Sahel13 — APPROVED** [review]
Yess test runs
